### PR TITLE
Add new buffer allocators to AFQMC

### PR DESCRIPTION
**Pull request by fdmalone**
## Proposed changes
The code now uses global stack allocators,
so the temporary space used by all routines now comes from the same place.
In GPUs, this means that buffer space in different parts of the code does not compete
against each other for space now. Should be useful when running larger systems
or when running estimators that burn a lot of memory.

## What type(s) of changes does this code introduce?

- New feature

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

quartz (intel xeon) + lassen ( ~ summit). gcc 7.3.0 

## Checklist

- Yes. This PR is up to date with current the current state of 'develop'
- No. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/AFQMC/HamiltonianOperations/KP3IndexFactorization_batched.hpp [hunk 1]
```diff
@@ -13,8 +13,8 @@
 //    Lawrence Livermore National Laboratory 
 ////////////////////////////////////////////////////////////////////////////////
 
-#ifndef QMCPLUSPLUS_AFQMC_HAMILTONIANOPERATIONS_KP3INDEXFACTORIZATION_BATCHED_HPP
-#define QMCPLUSPLUS_AFQMC_HAMILTONIANOPERATIONS_KP3INDEXFACTORIZATION_BATCHED_HPP
+#ifndef QMCPLUSPLUS_AFQMC_HAMILTONIANOPERATIONS_KP3INDEXFACTORIZATION_BATCHED_OOC_HPP
```
Comment (fdmalone) [inline src/AFQMC/HamiltonianOperations/KP3IndexFactorization_batched.hpp:16]: why is _ooc in the header include guard?

Comment (mmorale3) [reply src/AFQMC/HamiltonianOperations/KP3IndexFactorization_batched.hpp:16]: this is an oversight

File: src/AFQMC/Memory/buffer_allocators.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            host_memory_resource{},
+            std::size_t(10*1024*1024),
+            device_constructor<char>{}
+        );
+//#if defined(ENABLE_ACCELERATOR)
+#if defined(ENABLE_CUDA)
+  std::shared_ptr<device_allocator_generator_type> device_buffer_generator{nullptr};
+#else  
+// on cpu build, device and host are the same memory space
+  std::shared_ptr<device_allocator_generator_type> device_buffer_generator=host_buffer_generator;
```
[the remark this answers is not in the document]
Comment (mmorale3) [reply src/AFQMC/Memory/buffer_allocators.cpp:32]: This is a multi step process, I haven't decided where I'm going to place this.
Right now I'm getting the code modified to use them.
This is a global memory resource, to be shared by the  entire code.
I might actually leave it as a global object, but I'm not sure yet.
I have taken precautions to properly construct/destroy them at the moment.

File: src/AFQMC/HamiltonianOperations/tests/test_hamiltonian_operations.cpp [hunk 3]
```diff
@@ -55,6 +56,9 @@ namespace qmcplusplus
 
 using namespace afqmc;
 
+namespace afqmc {
+extern std::shared_ptr<device_allocator_generator_type> device_buffer_generator;
```
[the remark this answers is not in the document]
Comment (mmorale3) [reply src/AFQMC/HamiltonianOperations/tests/test_hamiltonian_operations.cpp:60]: There is no documentation for any of this yet.

File: src/AFQMC/Memory/buffer_allocators.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/AFQMC/Memory/buffer_allocators.cpp:32]: I get that, but I think decision to scope and lifetime of a resource are just as important as code having "access" to said resource.  We are trying to write a maintainable worldclass application so this needs to be resolved before production relies on this and the pain of fixing this increases.  I've filed an issue on this against the 4.0.0 release.

File: src/AFQMC/HamiltonianOperations/tests/test_hamiltonian_operations.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/AFQMC/HamiltonianOperations/tests/test_hamiltonian_operations.cpp:60]: It is a really good idea to write at little commentary at the moment of inception for "interesting" design decisions.  You may never be back to the code again.

**Comment by ye-luo** [comment]
@mmorale3 could you please update the PR title and description to mention this is specific to AFQMC.

My experience on the real space shows that in some aspects using large buffer brings cool optimizations. On the other hand, breaking, fusing scopes can be problematic.

In the case of AFQMC, are those buffers used as scratch space very locally and closer to the accelerator calls although owned globally?

**Comment by ye-luo** [comment]
Okay to test

**Comment by mmorale3** [comment]
Ye, they are various built-in safeguards.
These are not general allocators (we use other allocators for more general scenarios), 
they MUST be used as a stack. There is no restriction on scope, they are meant to be used globally (by any part of the AFQMC code), as long as they follow a stack pattern. The point is to replace all local needs for temporary buffer space (typically within the scope of a routine that gets called repeatedly) with these ones. The main need for this is to eliminate the need for routines to keep their own local buffer space. For large jobs, these buffers start to fight against each other and limit the size of the job we can fit on the GPU. The largest job I can run right now is not limited by speed, it is limited by the 16GB limit. These buffers, together with more changes I'm pushing later in the week, will free-up about 50% of the memory usage in the gpus. 

- If the allocation/deallocation cycle doesn't follow a stack pattern, the allocation throws.
- When buffers are updated (to allocate the exact amount of memory needed during a cycle), they must be empty otherwise they throw.
- these are actually fallback allocators too, so if the memory buffer is used up, allocation falls back to a secondary memory resource.

These were carefully design to be safe.

**Comment by mmorale3** [comment]
Designed by Afredo I must say.

**Pull request merged**